### PR TITLE
Corelli powder calibration: Implement translational constraint

### PR DESCRIPTION
**Pull request by jmborr**
**Description of work.**
- [x] property in algorithm `CorelliPowderCalibrationCreate` to fix the Y-coordinate of the banks
- [x] minimizer options in `AlignComponents`. Default minimizer is now `differential_evolution`
- [x] minimizer options in `CorelliPowderCalibrationCreate`. Default minimizer is now `differential_evolution`
- [x] unit test
- [ ] release notes

**To test:**
Run the general case for the unit test in the workbench:
```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
from numpy.testing import assert_allclose
#
# SETUP
#
spacings_reference = [0.9179, 0.9600, 1.0451, 1.2458, 1.3576, 1.5677, 1.6374, 3.1353]  # silicon
# add one Gaussian peak for every reference d-spacing
peak_functions = list()
for spacing in spacings_reference:
    peak_function = f'name=Gaussian, PeakCentre={spacing}, Height={10 * np.sqrt(spacing)}, Sigma={0.003 * spacing}'
    peak_functions.append(peak_function)
function = ';'.join(peak_functions)
begin, end, bin_width = spacings_reference[0] - 0.5, spacings_reference[-1] + 0.5, 0.0001
# Single 10x10 rectangular detector, located 4m downstream the sample along the X-axis
# Each detector has the same histogram of intensities, showing eight Gaussian peaks with centers at the
# reference d-spacings
CreateSampleWorkspace(WorkspaceType='Histogram', Function='User Defined', UserDefinedFunction=function,
                      XUnit='dSpacing', XMin=begin, XMax=end, BinWidth=bin_width,
                      NumBanks=1, PixelSpacing=0.02, SourceDistanceFromSample=10.0, BankDistanceFromSample=4.0,
                      OutputWorkspace='test_workspace_dSpacing')
RotateInstrumentComponent(Workspace='test_workspace_dSpacing', ComponentName='bank1', X=0., Y=1., z=0.,
                          Angle=90, RelativeRotation=True)
MoveInstrumentComponent(Workspace='test_workspace_dSpacing', ComponentName='bank1', X=4.0, y=0.0, z=0.0,
                        RelativePosition=False)
# Eight peaks now in TOF. Only when the instrument is located 4m downstream along the X-axis will we obtain
# the correct d-Spacings if we convert back to dSpacings units. If we perturb the instrument and convert
# back to dSpacing units, we'll obtain eight peaks centered at d-spacings sligthly different than the
# reference values
ConvertUnits(InputWorkspace='test_workspace_dSpacing', Target='TOF', EMode='Elastic',
             OutputWorkspace='test_workspace_TOF')
Rebin(InputWorkspace='test_workspace_TOF', Params=[300, 1.0, 16666.7], OutputWorkspace='test_workspace_TOF')
ConvertUnits(InputWorkspace='test_workspace_TOF', Target='dSpacing', EMode='Elastic',
             OutputWorkspace='test_workspace_dSpacing')
#
# PERTURB BANK POSITION AND ORIENTATION
#
CloneWorkspace(InputWorkspace='test_workspace_TOF', OutputWorkspace='perturbed')
MoveInstrumentComponent(Workspace='perturbed', ComponentName='bank1', X=0.02, y=0.005, z=0.005,
                        RelativePosition=True)
RotateInstrumentComponent(Workspace='perturbed', ComponentName='bank1', X=0, Y=0, z=1, Angle=5,
                          RelativeRotation=True)
ConvertUnits(InputWorkspace='perturbed', Target='dSpacing', EMode='Elastic', OutputWorkspace='perturbed_dS')              
#
# RUN CALIBRATION
#
CloneWorkspace(InputWorkspace='perturbed', OutputWorkspace='perturbed_calibrated')
CorelliPowderCalibrationCreate(
        InputWorkspace='perturbed_calibrated', OutputWorkspacesPrefix='cal_',
        TofBinning=[300, 1.0, 16666.7], PeakPositions=spacings_reference, SourceToSampleDistance=10.0,
        FixY=False, ComponentList='bank1', ComponentMaxTranslation=0.03, ComponentMaxRotation=6,
        Minimizer='differential_evolution', MaxIterations=10)
#
# COMPARE SPECTRA TO REFERENCE
#
ConvertUnits(InputWorkspace='perturbed_calibrated', Target='dSpacing', EMode='Elastic', OutputWorkspace='perturbed_calibrated_dS')
results = CompareWorkspaces(Workspace1='test_workspace_dSpacing', Workspace2='perturbed_calibrated_dS', Tolerance=0.001,
                            CheckInstrument=False)
print(results.Result)
```

The result of `print(results.Result)` should be `True`. You can also compare the spectra of the reference workspace `test_workspace_dSpacing` against the spectra of the perturbed instrument before ( workspace `perturbed_dS`) and after (workspace `perturbed_calibrated_dS`) calibration.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/plugins/algorithms/AlignComponents.py [hunk 1]
```diff
@@ -499,7 +515,641 @@ def _extract_tofs(self, table_tofs: TableWorkspace) -> np.ndarray:
...
+        #         ax1 = quat[2] / scale
+        #         ax2 = quat[3] / scale
+        #         return deg, ax0, ax1, ax2
+        #
+        #
+        # try:
+        #     from scipy.optimize import minimize, OptimizeResult
+        #     AlgorithmFactory.subscribe(AlignComponents)
+        # except ImportError:
+        #     logger.debug('Failed to subscribe algorithm AlignComponets; cannot import minimize from scipy.optimize')eak centers (d-spacing units)
```
Comment (KedoKudo) [inline Framework/PythonInterface/plugins/algorithms/AlignComponents.py:1152]: Is there any particular reason we want to keep the outdated code as comments?

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/CorelliPowderCalibrationCreateTest.py [hunk 2]
```diff
@@ -5,70 +5,134 @@
...
+                AdjustSource=False, FixY=False, ComponentList='bank1', ComponentMaxTranslation=0.2,
+                ComponentMaxRotation=10)
         except RuntimeError as error:
             assert 'Some invalid Properties found' in str(error)
 
-        # The calibration algorithm will attempt to correct the position and orientation of the bank so that peak
-        # centers for all detectors in the bank (not just detector-ID 155) approach our reference values. As
-        # a result, the final position and orientation is not exactly perpendicular to the X-axis and positioned
-        # five meters away from the sample.
+    @unittest.skip("causes surpassing the timeout in the Jenkins servers")
```
Comment (KedoKudo) [inline Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/CorelliPowderCalibrationCreateTest.py:92]: Running the full unit test (comment out the `@unittest.skip`) takes about 27 sec on the analysis cluster, which is way below the timeout limit on build server (5 min).  Am I missing something here?

**Review by KedoKudo — COMMENTED** [review]
The PR is built and tested (using the provided the testing script) on analysis cluster.  I'm happy to approve this PR once the release note is updated.

File: Framework/PythonInterface/plugins/algorithms/AlignComponents.py (hunk 1 above)
Comment (jmborr) [reply Framework/PythonInterface/plugins/algorithms/AlignComponents.py:1152]: This is some legacy "code", by Ross, I believe. I guess we can part with it at this time

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/CorelliPowderCalibrationCreateTest.py (hunk 2 above)
Comment (jmborr) [reply Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/CorelliPowderCalibrationCreateTest.py:92]: I think the build servers are not as quick as the analysis cluster. The first time I sent this PR, the test timed out at 300 seconds. Go figure!

**Comment by jmborr** [comment]
> The PR is built and tested (using the provided the testing script) on analysis cluster. I'm happy to approve this PR once the release note is updated.

I'm going to add the release notes, and cross my fingers there won't be any pesky conflicts

File: Framework/PythonInterface/plugins/algorithms/AlignComponents.py (hunk 1 above)
Comment (jmborr) [reply Framework/PythonInterface/plugins/algorithms/AlignComponents.py:1152]: Actually, it's me who did this...yesterday!
I have no clue how this happened.

**Review by KedoKudo — APPROVED** [review]

**Pull request merged**